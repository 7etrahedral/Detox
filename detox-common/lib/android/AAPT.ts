const _ = require("lodash");
const exec = require("child-process-promise").exec;
const Environment = require("../utils/environment");
const path = require("path");
const fsext = require("../utils/fsext");

export class AAPT {
  private aaptBin: string | null = null;

  public async _prepare() {
    if (!this.aaptBin) {
      const sdkPath = Environment.getAndroidSDKPath();
      const buildToolsDirs = await fsext.getDirectories(path.join(sdkPath, "build-tools"));
      const latestBuildToolsVersion = _.last(buildToolsDirs);
      this.aaptBin = path.join(sdkPath, "build-tools", latestBuildToolsVersion, "aapt");
    }
  }

  public async getPackageName(apkPath: string) {
    await this._prepare();

    const process = await exec(`${this.aaptBin} dump badging "${apkPath}"`);
    const packageName = new RegExp(/package: name='([^']+)'/g).exec(process.stdout);

    if (!packageName) {
      throw new Error("packageName is null");
    }

    return packageName[1];
  }
}

export default AAPT;
