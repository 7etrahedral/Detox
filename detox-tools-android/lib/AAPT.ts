import _ from 'lodash';
import fs from 'fs';
import {exec} from 'child-process-promise';
import {DetoxRuntimeError} from "./errors/DetoxRuntimeError";
import {DetoxChildProcessError} from "./errors/DetoxChildProcessError";
import {isChildProcessError} from "./utils/isChildProcessError";

export type AAPTOptions = {
    aaptBin: string;
};

export class AAPT {
    private static readonly PACKAGE_REGEXP = /package: name='([^']+)'/g;

    private readonly aaptBin: string;

    constructor(options: AAPTOptions) {
        this._assertOptionsAreNotEmpty(options);
        this._assertBinaryPathIsNotEmpty(options);
        this._assertBinaryPathExists(options);

        this.aaptBin = options.aaptBin;
    }

    /**
     * @see {@link https://stackoverflow.com/a/6289168}
     */
    public async readPackageNameFromAPK(apkPath: string): Promise<string> {
        this._assertApkPathIsNotEmpty(apkPath);

        const aapt = this.aaptBin;
        const command = `${aapt} dump badging "${apkPath}"`;

        const process = await exec(command).catch(e => e);
        if (_.isError(process)) {
            this._rethrowChildProcessError(process, command);
        }

        const packageName = AAPT.PACKAGE_REGEXP.exec(process.stdout);
        if (!packageName) {
            throw new DetoxChildProcessError({
                message: 'could not find the package name in aapt stdout',
                command,
                stdout: process.stdout,
                stderr: process.stderr,
            });
        }

        return packageName[1];
    }

    private _assertOptionsAreNotEmpty(options: AAPTOptions): void | never {
        if (!options) {
            throw new DetoxRuntimeError({
                message: `${AAPT.name}.constructor() did not receive any options`,
                hint: `Make sure you are passing non-empty options to new ${AAPT.name}(options)`,
            });
        }
    }

    private _assertBinaryPathIsNotEmpty(options: AAPTOptions): void | never {
        if (!options.aaptBin) {
            throw new DetoxRuntimeError({
                message: `${AAPT.name}.constructor(options) did not get binary path of *aapt*`,
                hint: `Options passed to constructor were: ${JSON.stringify(options)}`,
            });
        }
    }

    private _assertBinaryPathExists(options: AAPTOptions): void | never {
        if (!fs.existsSync(options.aaptBin)) {
            throw new DetoxRuntimeError({
                message: `*aapt* binary was not found`,
                hint: `Make sure it exists at path: ${options.aaptBin}`,
            });
        }
    }

    private _assertApkPathIsNotEmpty(apkPath: string): void | never {
        if (!apkPath) {
            throw new DetoxRuntimeError({
                message: 'No path to APK file is specified',
                hint: 'Make sure that AAPT.readPackageNameFromAPK(apkPath) gets valid path.'
            });
        }
    }

    private _rethrowChildProcessError(error: Error, command: string): void {
        /* istanbul ignore next */
        if (!isChildProcessError(error)) { // NOTE: this is a currently impossible case, but let's leave this check
            throw new DetoxChildProcessError({
                message: error.message,
                command,
                error,
            });
        }

        throw new DetoxChildProcessError({
            command,
            code: error.code,
            stdout: error.stdout,
            stderr: error.stderr,
        });
    }
}

