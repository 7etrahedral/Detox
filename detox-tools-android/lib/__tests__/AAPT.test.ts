import path from "path";
import {AAPT, AAPTOptions} from "../AAPT";
import jc from "../utils/test/JestErrorCaster";

const AAPT_HOME = path.join(path.relative('.', __dirname), "executables", "aapt");

describe(AAPT, () => {
    describe("constructor", () => {
        it("should throw when no args given", () => {
            const options: any = void 0;
            expect(jc.wrap(() => new AAPT(options))).toThrowErrorMatchingSnapshot();
        });

        it("should throw when aapt binary path is not specified", () => {
            expect(jc.wrap(() => new AAPT({ aaptBin: "" }))).toThrowErrorMatchingSnapshot();
        });

        it("should throw when aapt binary path does not exist", () => {
            const options: AAPTOptions = {
                aaptBin: "/tmp/unexisting_aapt_path",
            };

            expect(jc.wrap(() => new AAPT(options))).toThrowErrorMatchingSnapshot();
        });
    });

    describe("methods", () => {
        const _AAPT = AAPT.prototype;
        let aapt: AAPT;

        describe(_AAPT.readPackageNameFromAPK, () => {
            const apkPath = "path/to/file.apk";

            describe("when aapt executable works but prints non-intelligible output", () => {
                beforeEach(() => aapt = new AAPT({aaptBin: path.join(AAPT_HOME, "aapt-dumb-badging-weird")}));

                it("should throw exception with logs and debug information", async () => {
                    try {
                        await aapt.readPackageNameFromAPK(apkPath);
                    } catch (e) {
                        expect(jc.toSimpleError(e)).toMatchSnapshot();
                    }
                });
            });

            describe("when aapt executable works as expected", () => {
                beforeEach(() => aapt = new AAPT({aaptBin: path.join(AAPT_HOME, "aapt-dumb-badging-ok")}));

                it("should return package name it parses in stdout", async () => {
                    const packageName = await aapt.readPackageNameFromAPK(apkPath);
                    expect(packageName).toMatchSnapshot();
                });

                it("should throw if empty package name is given", async () => {
                    try {
                        await aapt.readPackageNameFromAPK("");
                    } catch (e) {
                        expect(jc.toSimpleError(e)).toMatchSnapshot();
                    }
                });
            });

            describe("when aapt executable exits with an error code", () => {
                beforeEach(() => aapt = new AAPT({aaptBin: path.join(AAPT_HOME, "aapt-failure")}));

                it("should throw an error with a verbose debug information", async () => {
                    try {
                        await aapt.readPackageNameFromAPK(apkPath);
                    } catch (e) {
                        expect(jc.toSimpleError(e)).toMatchSnapshot();
                    }
                });
            });
        });
    });

});
