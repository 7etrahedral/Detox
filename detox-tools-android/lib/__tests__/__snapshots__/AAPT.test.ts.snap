// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AAPT constructor should throw when aapt binary path does not exist 1`] = `
"*aapt* binary was not found
HINT: Make sure it exists at path: /tmp/unexisting_aapt_path"
`;

exports[`AAPT constructor should throw when aapt binary path is not specified 1`] = `
"AAPT.constructor(options) did not get binary path of *aapt*
HINT: Options passed to constructor were: {\\"aaptBin\\":\\"\\"}"
`;

exports[`AAPT constructor should throw when no args given 1`] = `
"AAPT.constructor() did not receive any options
HINT: Make sure you are passing non-empty options to new AAPT(options)"
`;

exports[`AAPT methods readPackageNameFromAPK when aapt executable exits with an error code should throw an error with a verbose debug information 1`] = `
[Error: Command failed with an error code (255)
To reproduce the error, run the following command:
lib/__tests__/executables/aapt/aapt-failure dump badging "path/to/file.apk"
You might find these dumps from stdout and stderr helpful. See below:
=====================================================================
starting...

aapt failed with an unknown error
]
`;

exports[`AAPT methods readPackageNameFromAPK when aapt executable works as expected should return package name it parses in stdout 1`] = `"com.wix.detox.test"`;

exports[`AAPT methods readPackageNameFromAPK when aapt executable works as expected should throw if empty package name is given 1`] = `
[Error: No path to APK file is specified
HINT: Make sure that AAPT.readPackageNameFromAPK(apkPath) gets valid path.]
`;

exports[`AAPT methods readPackageNameFromAPK when aapt executable works but prints non-intelligible output should throw exception with logs and debug information 1`] = `
[Error: could not find the package name in aapt stdout
To reproduce the error, run the following command:
lib/__tests__/executables/aapt/aapt-dumb-badging-weird dump badging "path/to/file.apk"
You might find these dumps from stdout and stderr helpful. See below:
=====================================================================
package: name="com.wix.detox.test" versionCode="1"
NOTE: regexp should fail on the line above, because package name is in double quotes

WARN: this warning should get into exception message in test
]
`;
